src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs

```diff
@@ -0,0 +1,262 @@
+        private LocalFuncAssignmentResults _localFuncAssignmentResults;
+
+        /// <summary>
+        /// True if we're currently recording reads, rather than reporting
+        /// diagnostics.
+        /// </summary>
+        private bool _recordingReads = false;
+
+        private SmallDictionary<LocalFunctionSymbol, LocalFuncInfo> _readVars =
+            new SmallDictionary<LocalFunctionSymbol, LocalFuncInfo>();
+
+
+        protected class LocalFuncInfo
+        {
+            public bool IsDirty { get; set; }
+            public PooledHashSet<int> UsedVars =
+                PooledHashSet<int>.GetInstance();
+
+            public void Free()
+            {
+                UsedVars.Free();
+            }
+        }
+
+        protected class LocalFuncAssignmentResults
+        {
+            public SmallDictionary<LocalFunctionSymbol, LocalFuncInfo> AssignedVars { get; }
+            public ImmutableArray<VariableIdentifier> VariableSlots { get; }
+
+            public LocalFuncAssignmentResults(
+                SmallDictionary<LocalFunctionSymbol, LocalFuncInfo> assignedVars,
+                ImmutableArray<VariableIdentifier> variableSlots)
+            {
+                AssignedVars = assignedVars;
+                VariableSlots = variableSlots;
+            }
+
+            public void Free()
+            {
+                FreeLocalFuncInfos(AssignedVars.Values);
+            }
+        }
+
+        private static void FreeLocalFuncInfos<T>(T infos)
+            where T : IEnumerable<LocalFuncInfo>
+        {
+            foreach (var info in infos)
+            {
+                info.Free();
+            }
+        }
+
+        protected static bool HasDirtyLocalFunctions<T>(T localFuncInfos)
+            where T : IEnumerable<LocalFuncInfo>
+        {
+            foreach (var info in localFuncInfos)
+            {
+                if (info.IsDirty)
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        protected static void ClearDirtyBits<T>(T localFuncInfos)
+            where T : IEnumerable<LocalFuncInfo>
+        {
+            foreach (var info in localFuncInfos)
+            {
+                info.IsDirty = false;
+            }
+        }
+
+        /// <summary>
+        /// The results from running the assignment pass.
+        /// </summary>
+        protected virtual LocalFuncAssignmentResults GetResults() => null;
+
+        /// <summary>
+        /// Records all possibly-unassigned reads in all local
+        /// functions in the given block.
+        /// </summary>
+        private void RecordReadsInLocalFunctions(BoundBlock block)
+        {
+            if (block.LocalFunctions.IsDefaultOrEmpty)
+            {
+                return;
+            }
+
+
+            var recordingState = _recordingReads;
+            _recordingReads = true;
+
+            do
+            {
+                ClearDirtyBits(_readVars.Values);
```

Comment:
True, but I think that's OK. Since we process in a depth-first search manner, each visited block will only contain dirty local functions below it, so all dirty local functions will necessary be below the current block being visited.